### PR TITLE
FRP GUI proof of concept

### PR DESCRIPTION
**Pull request by bkase**
Proof of concept GUI driven by FRP signals ala Cycle.js

### CODE REVIEW

File: src/future_driver.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+use futures::task::Poll;
+use futures::Future;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+pub fn tick<F: Future>(f: Rc<RefCell<F>>) -> Poll<F::Output> {
+    let waker = futures_util::task::noop_waker();
+    // TODO: Is this safe?
```
Comment (cmr) [inline src/future_driver.rs]: no, I think you need `F: Future+Unpin`.

Comment (cmr) [reply src/future_driver.rs]: `Pin<T>` means that until the `T` is dropped it isn't ever moved -- crucially _not_ until the `Pin<T>` is dropped- once pinned, always pinned. And you can't control that here- depending on the other owners of the `Rc<_>`.

File: src/lib.rs [hunk 2]
```diff
@@ -73,13 +87,24 @@ pub fn run() -> Result<(), JsValue> {
...
+            *lock = i;
+        }
+
         // Stop after 500 frames
         if i > 500 {
             // Drop our handle to this closure so that it will get cleaned
             // up once we return.
             let _ = f.borrow_mut().take();
+            // TODO: Does this free the memory properly?
+            let _ = signal_future.borrow_mut();
```
Comment (cmr) [inline src/lib.rs]: calling borrow_mut and doing nothing is a no-op.

**Comment by bkase** [comment]
Thanks!

**Pull request merged**